### ISSUE TITLE
Comparing cron expressions based on their frequency 

### ISSUE THREAD

**Issue by sfatali** [jmrozanec/cron-utils#511]
Hi!
Would love to see a feature for comparing cron expressions based on their frequency. Would be useful e.g. when sorting a collection of cron expressions based on how frequent they fire up.

**Comment by jmrozanec** [comment]
@sfatali thank you for the suggestion! 😄 We are categorizing the request among features. If you have some time to contribute a solution, we will be glad to merge it and release a version with it. Thank you for helping us make cron-utils better! 😄

**Comment by lodenrogue** [comment]
I've been thinking through this one and I haven't been able to come up with a solution. I keep thinking of an approach where you can check how often one expression fires before another expression fires but that won't work because the first could stop at that moment and the second expression could fire more often than the first after that moment in time. We would need to calculate until the end of time with this approach.

I think a better solution might be to sort by frequency within a given time period. For example, frequency between a given start and end date. That way we could leverage the other solution I wrote.

Otherwise, do we have a way to know if a cron expression has looped around? Has it gone through it's whole execution cycle and there is no more triggering left to happen? For example, an expression that triggers every Monday at the 5th minute of every hour. Once it's past 11:05 pm on Monday it would loop around waiting until the next Monday.

**Comment by lodenrogue** [comment]
Thinking through it more there's probably a way to do some math and come up with a number to represent the frequency:

Take this expression for example:
```
*/4 7,15,21 * *
```

We could do something like:
```
(1/60)*(4/24)*(3/30.4375)*(1)*(1) = 0.000273785 times per minute
```

I'll just need some help figuring out how to apply this to what cron-utils has in place today. Also, someone who is more familiar with cron expressions and all their variations please tell me where this could go wrong.

**Comment by jmrozanec** [comment]
@lodenrogue for sure we need a timeframe to compute; otherwise, it could be infinite for recurring executions. Moreover, given an arbitrary timeframe, we need to make this computation efficient: e.g., if we have a cron expression that asks to fire certain action every second, then it would make more than two and a half million executions in a month (about 2.629.743 to be more precise). For greater intervals, this could get very difficult to manage. 
Perhaps we should think about how to split the problem into minor pieces, and check for executions in each of them (some kind of "execution search") for all possible occurrences in a given timeframe. Our existing heuristics are optimized for finding the next/last execution efficiently, but not to compute all possible occurrences in such a way. 
One idea could be some pattern matching with constraints: build some tree of datetimes matching the particular conditions of the cron, and then list all datetimes that match the given cron expression.

**Comment by jmrozanec** [comment]
@lodenrogue the feature was implemented with a custom Comparator implementation, which requires to define a date range to perform the comparison. Further enhancements are possible 😄 We will release the feature soon. Thank you for pushing the boundaries of cron-utils!

**Comment by jmrozanec** [comment]
@sfatali @lodenrogue, the abovementioned functionality was implemented and released in version 9.2.0. Thank you for helping us build a better cron-utils! 🚀 😎
